Proposed feature addition to git-parameter plugin for Jenkins

graeme-hill opened the pull request:
Hi.  In my fork of git-parameter I added an option called tagFilter that allows you to set the filter pattern for listing tags.  For example, if you're using tags of the form v:1.0.0.1234 to manage your releases and you don't want other tags to show up in the list, then you can set your tag filter to "v:*" and then only tags that have been intentionally setup to use the versioning convention will be displayed.

graeme-hill committed:
Added a new option for filtering tag list

The new parameter is called tagFilter.  Previously tags were always
listed with the command "git tag -l *".  Now the * is replaced with
whatever value is specified for tagFilter. If tagFilter is empty or
whitespace then * is used just to be explicit.

graeme-hill committed:
Added tagPartSplitter parameter for tag sorting

If you use a tag format like v_1.2.3.4 then lexicographical order does
not sort properly because a given version component could have different
numbers of digits.  To fix this sorting can be done with awareness of
the tokens that make up the tag. To break the tag into tokens you
specify a tagPartSplitter pattern that acts as a regular expression. If
the parameter is blank then a direct lexicographical comparison will be
used.

graeme-hill committed:
Changed tag map to make sorting work correctly

In the last commit I already implemented tag sorting with the tag part
splitter parameter, but it still wasn't actually showing up in the right
order in jenkins. The collection of tags is stored in a java map where
previously both the key and value were the name of the tag which caused
the order to get lost even though the tags were being inserted to the
map in the right order. Now the keys are just integers starting from 0
and the values are the tag names. It appears that jenkins sorts by tag
name because it seems to be showing up correctly now.

graeme-hill committed:
Failed changes to sort tags in UI

I made the displayed values start with incrementing integers (with
leading zeros) so that lexicographical order would work properly, but it
doesn't seem to even do that. Possibly the behaviour has to do with
HashMap having no controllable sort order.

lukanus commented:
Hey. Thanks for patch, but I would like to check it first, before I'll merge it. Sadly, I don't have much time lately, so, I'll do this in following week, if You don't mind :).

graeme-hill commented:
Yeah no rush.  I'm really new to Jenkins so you definitely should check out the code to make sure I didn't break anything :)

By the way I also tried to add a sorting parameter in commits c730019 and ce42b40 but on closer inspection there is a flaw in my logic and I also have not been able to control the order that elements get displayed in the actual selector. As far as I know the tagFilter parameter in commit 205d7bb works fine though.

graeme-hill committed:
Replaced tag splitter with smart sort param

If use smart number sort is enabled then consecutive digits in the tag
are treated as a single character.  This means that the following tags
would sort to this order:

v_1.1.1
v_1.1.2
v_1.1.10

with smart number sorting turned off they would end up in this order:

v_1.1.1
v_1.1.10
v_1.1.2

graeme-hill committed:
Fixed ordering of tags in select box

Revisions and tags were stored in HashMap objects which means they do
not really have an order. Changed HashMap to LinkedHashMap so now the
keys are ordered as they are insterted which allows control over sort
order that shows up in the Jenkins UI.

graeme-hill committed:
Created sortMode parameter

Instead of using a bunch of check boxes for enabling smart sort and
choosing ascending or descending sort order I created an enum with
values NONE, ASCENDING, DESCENDING, ASCENDING_SMART, and
DESCENDING_SMART.  There is now a combo box parameter where you can
choose one of the sort methods.

graeme-hill commented:
Okay, it seems to be finally working as I originally intended after commit 92af536.

ngiger commented:
Merge functionality (with some adaptions into https://github.com/jenkinsci/git-parameter-plugin.git) . Try version 0.3.2 to see whether it works for you.

graeme-hill commented:
Cool
